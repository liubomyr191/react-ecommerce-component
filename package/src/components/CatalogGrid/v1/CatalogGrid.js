import React, { Component } from "react";
import PropTypes from "prop-types";
import { ContainerQuery } from "react-container-query";
import styled from "styled-components";
import { withComponents } from "@reactioncommerce/components-context";
import { getFromTheme, CustomPropTypes, preventAccidentalDoubleClick } from "../../../utils";

const mdWidth = getFromTheme({}, "rui_md");
const containerQueries = {
  is2PerRowWidth: {
    minWidth: 450, // Min width that item w/ 2 badges renders appropriately
    maxWidth: 649
  },
  is3PerRowWidth: {
    minWidth: 650,
    maxWidth: mdWidth - 1
  },
  is4PerRowWidth: {
    minWidth: mdWidth
  }
};

const GridContainer = styled.div`
  display: flex;
  flex-wrap: wrap;
  box-sizing: border-box;
  width: 100%`;

const GridItem = styled.div`
  padding: 12px;
  flex-grow: 0;
  max-width: 100%;
  flex-basis: 100%;
  margin: 0;
  box-sizing: border-box;

  ${({ containerParams }) => {
    const { is2PerRowWidth, is3PerRowWidth, is4PerRowWidth } = containerParams;
    if (is2PerRowWidth) {
      return `
        max-width: 50%;
        flex-basis: 50%;
      `;
    } else if (is3PerRowWidth) {
      return `
        max-width: 33.33333%;
        flex-basis: 33.33333%;
      `;
    } else if (is4PerRowWidth) {
      return `
        max-width: 25%;
        flex-basis: 25%;
      `;
    }
    return "";
  }}
  `;

class CatalogGrid extends Component {
  static propTypes = {
    /**
     * Labels to use for the various badges. Refer to `BadgeOverlay`'s prop documentation.
     */
    badgeLabels: PropTypes.shape({
      BACKORDER: PropTypes.string,
      BESTSELLER: PropTypes.string,
      LOW_QUANTITY: PropTypes.string,
      SOLD_OUT: PropTypes.string,
      SALE: PropTypes.string
    }),
    /**
     * If you've set up a components context using @reactioncommerce/components-context
     * (recommended), then this prop will come from there automatically. If you have not
     * set up a components context or you want to override one of the components in a
     * single spot, you can pass in the components prop directly.
     */
    components: PropTypes.shape({
      CatalogGridItem: CustomPropTypes.component.isRequired
    }).isRequired,
    /**
     * Currency code to display the price for. Product must include a pricing object with the code in `product.pricing`
     */
    currencyCode: PropTypes.string,
    /**
     * The inital size the grid should render at. Use to set grid width during SSR.
     */
    initialSize: PropTypes.shape({
      height: PropTypes.number,
      width: PropTypes.number
    }),
    /**
     * Item click handler
     */
    onItemClick: PropTypes.func,
    /**
     * Image to display when product doesn't have a primary image
     */
    placeholderImageURL: PropTypes.string,
    /**
     * Products to display in the grid. Refer to `CatalogGridItem`'s documentation
     */
    products: PropTypes.arrayOf(PropTypes.object)
  };

  static defaultProps = {
    badgeLabels: null,
    currencyCode: "USD",
    initialSize: {
      width: 325
    },
    onItemClick() {},
    placeholderImageURL: "/resources/placeholder.gif",
    products: []
  };

  handleOnClick = preventAccidentalDoubleClick((event, product) => {
    this.props.onItemClick(event, product);
  });

  render() {
    const {
      badgeLabels,
      components: { CatalogGridItem },
      currencyCode,
      initialSize,
      onItemClick,
      placeholderImageURL,
      products
    } = this.props;

    const gridItemProps = {
      currencyCode,
      placeholderImageURL,
      onClick: onItemClick
    };

    if (badgeLabels) {
      gridItemProps.badgeLabels = badgeLabels;
    }

    return (
      <ContainerQuery query={containerQueries} initialSize={initialSize}>
        {(params) => (
          <GridContainer>
            {products.map((product, index) => (
              <GridItem containerParams={params} key={`grid-item-${index}`} {...this.props}>
                <CatalogGridItem product={product} {...gridItemProps} />
              </GridItem>
            ))}
          </GridContainer>
        )}
      </ContainerQuery>
    );
  }
}

export default withComponents(CatalogGrid);
