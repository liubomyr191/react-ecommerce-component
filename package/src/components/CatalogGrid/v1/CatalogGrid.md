### Overview
The `CatalogGrid` component creates a responsive grid that displays `CatalogGridItem`s.

#### Usage

##### Default - responsive grid
`CatalogGrid` automatically adjusts itself based on the width of its container. This ensures
that products always render appropriately regardless of where the grid is rendered.

```js noeditor
  // Import mock products behind the scenes
  products = require("./__mocks__/products").default; ""
```

```jsx
  <CatalogGrid products={products} onItemClick={(event) => event.preventDefault()} />
```

##### Fixed-width container, 1 product per row (325px width)

```jsx
  <div style={{ width: 325, borderWidth: 1, borderColor: "black", borderStyle: "solid", margin: "0 auto" }}>
    <CatalogGrid products={products} />
  </div>
```

##### Fixed-width container, 2 per row (450px width)

```jsx
  <div style={{ width: 450, borderWidth: 1, borderColor: "black", borderStyle: "solid", margin: "0 auto" }}>
    <CatalogGrid products={products} />
  </div>
```

##### Fixed-width container, 3 per row (650px width)

```jsx
  <div style={{ width: 650, borderWidth: 1, borderColor: "black", borderStyle: "solid", margin: "0 auto" }}>
    <CatalogGrid products={products} />
  </div>
```

##### Fixed-width container, 4 per row (960px width)

```jsx
  <div style={{ width: 960, borderWidth: 1, borderColor: "black", borderStyle: "solid", margin: "0 auto" }}>
    <CatalogGrid products={products} />
  </div>
```

##### Custom placeholder image
By default, `CatalogGrid` expects a placeholder image to be located at /resources/placeholder.gif.
Specify the `placeholderImageURL` to override:

```js noeditor
  // Clone products and remove images
  productsWithoutImages = products.slice().map((product) => {
    const clonedProduct = Object.assign({}, product);
    clonedProduct.primaryImage = null;
    return clonedProduct;
  });""
```

```jsx
  <CatalogGrid
    placeholderImageURL="/images/placeholder.gif"
    products={productsWithoutImages}
  />
```
