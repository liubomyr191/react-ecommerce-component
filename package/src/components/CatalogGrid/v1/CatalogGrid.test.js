import React from "react";
import renderer from "react-test-renderer";
import mockComponents from "../../../tests/mockComponents";
import CatalogGrid from "./CatalogGrid";
import mockProducts from "./__mocks__/products";

test("CatalogGrid default responsive snapshot", () => {
  const component = renderer.create((
    <CatalogGrid
      components={mockComponents}
      products={mockProducts}
    />
  ));

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("CatalogGrid 325px fixed-width snapshot", () => {
  const component = renderer.create((
    <div style={{ width: 325, borderWidth: 1, borderColor: "black", borderStyle: "solid", margin: "0 auto" }}>
      <CatalogGrid
        components={mockComponents}
        products={mockProducts}
      />
    </div>
  ));

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("CatalogGrid 450px fixed-width snapshot", () => {
  const component = renderer.create((
    <div style={{ width: 450, borderWidth: 1, borderColor: "black", borderStyle: "solid", margin: "0 auto" }}>
      <CatalogGrid
        components={mockComponents}
        products={mockProducts}
      />
    </div>
  ));

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("CatalogGrid 650px fixed-width snapshot", () => {
  const component = renderer.create((
    <div style={{ width: 650, borderWidth: 1, borderColor: "black", borderStyle: "solid", margin: "0 auto" }}>
      <CatalogGrid
        components={mockComponents}
        products={mockProducts}
      />
    </div>
  ));

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("CatalogGrid 960px fixed-width snapshot", () => {
  const component = renderer.create((
    <div style={{ width: 960, borderWidth: 1, borderColor: "black", borderStyle: "solid", margin: "0 auto" }}>
      <CatalogGrid
        components={mockComponents}
        products={mockProducts}
      />
    </div>
  ));

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("CatalogGrid custom placeholder image snapshot", () => {
  const component = renderer.create((
    <CatalogGrid
      components={mockComponents}
      products={mockProducts}
      placeholderImageURL={"/path/to/custom/placeholder.png"}
    />
  ));

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
