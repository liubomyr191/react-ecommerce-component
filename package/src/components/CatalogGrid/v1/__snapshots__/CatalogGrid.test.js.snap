// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CatalogGrid basic snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  box-sizing: border-box;
  width: calc(100% + 24px);
  margin: -12px;
}

.c1 {
  padding: 12px;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  max-width: 100%;
  -webkit-flex-basis: 100%;
  -ms-flex-preferred-size: 100%;
  flex-basis: 100%;
  margin: 0;
  box-sizing: border-box;
}

@media (min-width:600px) {
  .c1 {
    -webkit-box-flex: 0;
    -webkit-flex-grow: 0;
    -ms-flex-positive: 0;
    flex-grow: 0;
    max-width: 33.33333%;
    -webkit-flex-basis: 33.33333%;
    -ms-flex-preferred-size: 33.33333%;
    flex-basis: 33.33333%;
  }
}

@media (min-width:960px) {
  .c1 {
    -webkit-box-flex: 0;
    -webkit-flex-grow: 0;
    -ms-flex-positive: 0;
    flex-grow: 0;
    max-width: 25%;
    -webkit-flex-basis: 25%;
    -ms-flex-preferred-size: 25%;
    flex-basis: 25%;
  }
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    CatalogGridItem(undefined)
  </div>
  <div
    className="c1"
  >
    CatalogGridItem(undefined)
  </div>
  <div
    className="c1"
  >
    CatalogGridItem(undefined)
  </div>
  <div
    className="c1"
  >
    CatalogGridItem(undefined)
  </div>
</div>
`;

exports[`CatalogGrid custom min-width breakpoints snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  box-sizing: border-box;
  width: calc(100% + 24px);
  margin: -12px;
}

.c1 {
  padding: 12px;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  max-width: 100%;
  -webkit-flex-basis: 100%;
  -ms-flex-preferred-size: 100%;
  flex-basis: 100%;
  margin: 0;
  box-sizing: border-box;
}

@media (min-width:960px) {
  .c1 {
    -webkit-box-flex: 0;
    -webkit-flex-grow: 0;
    -ms-flex-positive: 0;
    flex-grow: 0;
    max-width: 33.33333%;
    -webkit-flex-basis: 33.33333%;
    -ms-flex-preferred-size: 33.33333%;
    flex-basis: 33.33333%;
  }
}

@media (min-width:1280px) {
  .c1 {
    -webkit-box-flex: 0;
    -webkit-flex-grow: 0;
    -ms-flex-positive: 0;
    flex-grow: 0;
    max-width: 25%;
    -webkit-flex-basis: 25%;
    -ms-flex-preferred-size: 25%;
    flex-basis: 25%;
  }
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    CatalogGridItem(undefined)
  </div>
  <div
    className="c1"
  >
    CatalogGridItem(undefined)
  </div>
  <div
    className="c1"
  >
    CatalogGridItem(undefined)
  </div>
  <div
    className="c1"
  >
    CatalogGridItem(undefined)
  </div>
</div>
`;

exports[`CatalogGrid custom placeholder image snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  box-sizing: border-box;
  width: calc(100% + 24px);
  margin: -12px;
}

.c1 {
  padding: 12px;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  max-width: 100%;
  -webkit-flex-basis: 100%;
  -ms-flex-preferred-size: 100%;
  flex-basis: 100%;
  margin: 0;
  box-sizing: border-box;
}

@media (min-width:600px) {
  .c1 {
    -webkit-box-flex: 0;
    -webkit-flex-grow: 0;
    -ms-flex-positive: 0;
    flex-grow: 0;
    max-width: 33.33333%;
    -webkit-flex-basis: 33.33333%;
    -ms-flex-preferred-size: 33.33333%;
    flex-basis: 33.33333%;
  }
}

@media (min-width:960px) {
  .c1 {
    -webkit-box-flex: 0;
    -webkit-flex-grow: 0;
    -ms-flex-positive: 0;
    flex-grow: 0;
    max-width: 25%;
    -webkit-flex-basis: 25%;
    -ms-flex-preferred-size: 25%;
    flex-basis: 25%;
  }
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    CatalogGridItem(undefined)
  </div>
  <div
    className="c1"
  >
    CatalogGridItem(undefined)
  </div>
  <div
    className="c1"
  >
    CatalogGridItem(undefined)
  </div>
  <div
    className="c1"
  >
    CatalogGridItem(undefined)
  </div>
</div>
`;
